components/sic_image_engine/src/operations/threshold.rs

```diff
@@ -0,0 +1,46 @@
+use crate::errors::SicImageEngineError;
+use crate::operations::ImageOperation;
+use rayon::iter::{IntoParallelRefMutIterator, ParallelIterator};
+use sic_core::{
+    image::{self, DynamicImage, RgbaImage},
+    SicImage,
+};
+
+pub struct Threshold;
+
+impl Threshold {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl ImageOperation for Threshold {
+    fn apply_operation(&self, image: &mut SicImage) -> Result<(), SicImageEngineError> {
+        match image {
+            SicImage::Static(image) => *image = threshold_image(image),
+            SicImage::Animated(image) => threshold_animated_image(image.frames_mut()),
+        }
+
+        Ok(())
+    }
+}
+
+fn threshold_image(img: &mut DynamicImage) -> DynamicImage {
+    let gray_image = img.to_luma8();
+    let best_threshold = imageproc::contrast::otsu_level(&gray_image);
+    let out = imageproc::contrast::threshold(&gray_image, best_threshold);
+    DynamicImage::ImageRgba8(DynamicImage::ImageLuma8(out).into_rgba8())
```

Comment:
It's probably not necessary to convert back to `rgba8` with `into_rgba8`. Since `SicImage` takes a `DynamicImage`, you could keep the buffer as `DynamicImage::ImageLuma8` in this case.

Reply:
Done thank you!